pkg/http/sender.go

```diff
@@ -0,0 +1,605 @@
+		case "tls_verify":
+			switch v {
+			case "on":
+				opts = append(opts, WithTls())
+			case "unsafe_off":
+				opts = append(opts, WithTlsInsecureSkipVerify())
+			default:
+				return nil, conf.NewConfigStrParseError("invalid tls_verify value, %q is not 'on' or 'unsafe_off", v)
+			}
+		case "tls_roots":
+			return nil, conf.NewConfigStrParseError("tls_roots is not available in the go client")
+		case "tls_roots_password":
+			return nil, conf.NewConfigStrParseError("tls_roots_password is not available in the go client")
+		default:
+			return nil, conf.NewConfigStrParseError("unsupported option %q", k)
+		}
+
+	}
+	return NewLineSender(opts...)
+}
+
+// Flush flushes the accumulated messages to the underlying HTTP
+// client. Should be called periodically to make sure that
+// all messages are sent to the server.
+//
+// For optimal performance, this method should not be called after
+// each ILP message. Instead, the messages should be written in
+// batches followed by a Flush call. Optimal batch size may vary
+// from 100 to 1,000 messages depending on the message size and
+// configured buffer capacity.
+func (s *LineSender) Flush(ctx context.Context) error {
+	var (
+		req           *http.Request
+		retryInterval time.Duration
+
+		maxRetryInterval = time.Second
+	)
+
+	err := s.LastErr()
+	s.ClearLastErr()
+	if err != nil {
+		s.DiscardPendingMsg()
+		return err
+	}
+	if s.HasTable() {
+		s.DiscardPendingMsg()
+		return errors.New("pending ILP message must be finalized with At or AtNow before calling Flush")
+	}
+
+	uri := "http"
+	if s.tlsMode > 0 {
+		uri += "s"
+	}
+	uri += fmt.Sprintf("://%s/write", s.address)
+
+	// timeout = ( request.len() / min_throughput ) + grace
+	// Conversion from int to time.Duration is in milliseconds
+	timeout := time.Duration(s.Len()/s.minThroughputBytesPerSecond)*time.Second + s.graceTimeout
+	reqCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	req, err = http.NewRequestWithContext(
+		reqCtx,
+		http.MethodPost,
+		uri,
+		s,
+	)
+	if err != nil {
+		return err
+	}
+
+	if s.user != "" && s.pass != "" {
+		req.SetBasicAuth(s.user, s.pass)
+	}
+
+	if s.token != "" {
+		if req.Header.Get("Authorization") != "" {
+			return errors.New("cannot use both Basic and Token Authorization")
+		}
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", s.token))
+	}
+
+	resp, err := s.client.Do(req)
+	if err == nil {
+		defer resp.Body.Close()
+
+		// If the requests succeeds with a non-error status code, flush has succeeded
+		if !isRetryableError(resp.StatusCode) {
+			// bytes.Buffer grows as 2*cap+n, so we use 3x as the threshold.
+			if s.Cap() > 3*s.BufCap {
+				// Shrink the buffer back to desired capacity.
+				s.Buffer.Buffer = *bytes.NewBuffer(make([]byte, s.InitBufSizeBytes, s.BufCap))
+			}
+			return nil
+		}
+		// Otherwise, we will retry sending the request until the retry timeout is breached
+		err = fmt.Errorf("Non-OK Status Code %d: %s", resp.StatusCode, resp.Status)
```

Comment:
We should parse error messages sent from the server and propagate them to the user. See `JsonErrorParser` in the Java client and its usage:
https://github.com/questdb/questdb/blob/713847e72f0c6977b61bf3001094dc60c6939c74/core/src/main/java/io/questdb/cutlass/line/http/LineHttpSender.java#L512